fedimint-api/src/db/mod.rs

```diff
@@ -247,6 +261,77 @@ impl Drop for CommitTracker {
     }
 }
 
+/// ModuleDatabaseTransaction is a wrapper around IsolatedDatabaseTransaction that
+/// consumes an existing DatabaseTransaction. This allows entire Databases to be
+/// isolated and calling begin_transaction will always produce a ModuleDatabaseTransaction,
+/// which is isolated from other modules by prepending a prefix to each key.
+struct ModuleDatabaseTransaction<'a> {
+    dbtx: DatabaseTransaction<'a>,
+    prefix: ModuleInstanceId,
+}
+
+impl<'a> ModuleDatabaseTransaction<'a> {
+    pub fn new(
+        dbtx: DatabaseTransaction<'a>,
+        prefix: ModuleInstanceId,
+    ) -> ModuleDatabaseTransaction<'a> {
+        ModuleDatabaseTransaction { dbtx, prefix }
+    }
+}
+
+#[async_trait]
+impl<'a> IDatabaseTransaction<'a> for ModuleDatabaseTransaction<'a> {
+    async fn raw_insert_bytes(&mut self, key: &[u8], value: Vec<u8>) -> Result<Option<Vec<u8>>> {
+        (&mut self.dbtx)
+            .with_module_prefix(self.prefix)
+            .raw_insert_bytes(key, value)
+            .await
+    }
+
+    async fn raw_get_bytes(&mut self, key: &[u8]) -> Result<Option<Vec<u8>>> {
+        (&mut self.dbtx)
+            .with_module_prefix(self.prefix)
+            .raw_get_bytes(key)
+            .await
+    }
+
+    async fn raw_remove_entry(&mut self, key: &[u8]) -> Result<Option<Vec<u8>>> {
+        (&mut self.dbtx)
+            .with_module_prefix(self.prefix)
+            .raw_remove_entry(key)
+            .await
+    }
+
+    async fn raw_find_by_prefix(&mut self, key_prefix: &[u8]) -> PrefixIter<'_> {
+        Box::new(
+            self.dbtx
+                .with_module_prefix(self.prefix)
+                .raw_find_by_prefix(key_prefix)
+                .await
+                .collect::<Vec<_>>()
```

Comment:
I don't think this is the greatest, but we'll probably need to wait to fix this until https://github.com/fedimint/fedimint/issues/1322 is fixed.

Reply:
We could already implement it (probably iface + dyn-newtype, again), and eventually we  will just turn it from `impl Iterator ...` to `impl Stream ...`.

Reply:
I'll leave for a follow up to keep this PR focused